Feature/error levels

meganrm opened the pull request:
Problem
=======
closes #219 
closes #201 

Solution
========
Added a level to front end error
Made all emitted errors FrontEndErrors
Left htmlData as a field because of how it's being used on the front end.

## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)
* This change requires updated or new tests

Steps to Verify:
----------------
I pushed my local build to the website repo if you want to test out these changes
2. View this [diff](https://github.com/allen-cell-animated/simularium-website/compare/local-viewer-install?expand=1)
3. Copy changes locally, plus install your viewer locally
4. Drag and drop the actin file without any geometry. 
5. You should never see the panel flash into error state, it goes from loading to success.

toloudis commented on examples/Viewer.tsx:
```diff
@@ -134,6 +136,16 @@ class Viewer extends React.Component<{}, ViewerState> {
+        } else if (error.level === ErrorLevel.WARNING) {
+            window.alert(`User warning, but not terrible: ${error.htmlData}`);
+        } else {
+            console.log(`Just for your info: ${error.htmlData}`);
```
here's the case I brought up where you are `console.log`ing html and it would make sense to have a non-html message

toloudis replied:
i wonder if there is a utility function to just remove any `<tags>` from a string

toloudis replied:
... or maybe the console.log should log error.message?

meganrm replied:
I was a bit mistaken when we talked this morning: the errors always have a message, they may also have htmlData. I updated the example viewer to log just the message

schoinh reviewed (commented):
When I search for `throw new Error` in this branch, I still see a lot of them (like [here](https://github.com/allen-cell-animated/simularium-viewer/blob/8d433af321fac55e8981becd285c02e4248a7324/src/simularium/RemoteSimulator.ts#L331)). Just want to check if you made sure these are all caught and emitted as a FrontEndError when needed? I wish it were easier to tell from our code

meganrm commented on src/controller/index.ts:
```diff
@@ -46,7 +47,7 @@ export default class SimulariumController {
     public handleTrajectoryInfo: (TrajectoryFileInfo) => void;
     public postConnect: () => void;
-    public onError?: (errorMessage: string) => void;
+    public onError?: (errorMessage: FrontEndError) => void;
```
Yeah I can change those, thanks

meganrm commented:
> When I search for `throw new Error` in this branch, I still see a lot of them (like [here](https://github.com/allen-cell-animated/simularium-viewer/blob/8d433af321fac55e8981becd285c02e4248a7324/src/simularium/RemoteSimulator.ts#L331)). Just want to check if you made sure these are all caught and emitted as a FrontEndError when needed? I wish it were easier to tell from our code

I went through and added comments to the code. Mostly those were all Errors being created within Promises or try catch blocks, that would then be passed up by the function that called it. I didn't want to make `FrontEndErrors` except for the places were it's truly the error being sent to the front end. If you're throwing an error within a function to prematurely exit, that should just be a normal Error. 

Thanks for asking though, because I did find one I missed, and I updated the logs on the example viewer so if an error gets sent in that doesn't have a level (ie is not a FrontEndError) we will hopefully be aware of it. 

There was one place where we are catching and error and logging it that might be better as a FrontEndError.

schoinh reviewed (approved):
Wow thank you for hunting down and fixing up all the errors, this is so much better than before
